pkg/providers/zone/zone.go

```diff
@@ -0,0 +1,157 @@
+/*
+Portions Copyright (c) Microsoft Corporation.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package zone
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armsubscriptions"
+	"github.com/samber/lo"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const maxFailures = 10
+
+// SubscriptionsAPI defines the interface for Azure Subscriptions client operations
+type SubscriptionsAPI interface {
+	NewListLocationsPager(
+		subscriptionID string,
+		options *armsubscriptions.ClientListLocationsOptions,
+	) *runtime.Pager[armsubscriptions.ClientListLocationsResponse]
+}
+
+// Provider handles zone support detection for Azure regions
+type Provider struct {
+	subscriptionsAPI SubscriptionsAPI
+	subscriptionID   string
+
+	// Cached zone support data - maps region name to zone support boolean
+	zoneSupport map[string]bool
+	hasLoaded   bool
+	// failures is the number of times loading zone support from the Azure API has failed
+	failures int
+	mu       sync.Mutex
+}
+
+// NewProvider creates a new zone provider
+func NewProvider(subscriptionsAPI SubscriptionsAPI, subscriptionID string) *Provider {
+	return &Provider{
+		subscriptionsAPI: subscriptionsAPI,
+		subscriptionID:   subscriptionID,
+		zoneSupport:      lo.Assign(fallbackZonalRegions), // deepcopy
+
+	}
+}
+
+// SupportsZones returns true if the given region supports availability zones
+func (p *Provider) SupportsZones(ctx context.Context, region string) bool {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	if !p.hasLoaded && p.failures < maxFailures {
+		// Try to load zone support data from Azure API
+		if err := p.loadFromAzure(ctx); err != nil {
+			p.failures++
+			log.FromContext(ctx).Error(err, "failed to load zone support from Azure API, falling back to hardcoded list")
+		} else {
+			p.hasLoaded = true
+		}
+	}
+
+	return p.zoneSupport[region] // if cache doesn't have our region, assume no zone support
+}
+
+// loadFromAzure discovers zone support by calling Azure Subscriptions API
+func (p *Provider) loadFromAzure(ctx context.Context) error {
```

Comment:
We could try to load this from Azure on another thread right from the start... which might be better?

Reply:
When is this provider called? If it is in the path of scale up - then loading on startup (and I am not even sure it needs to be on a different thread) would be better, as it would shave a bit of latency from the first scaleup. If it is not in the path of scaleup (e.g. maybe loaded when instance types are computed?) - then it does not matter, I think ...

Reply:
Discussed, agreed how it is now is best. Left a comment with the reasoning in this method.